### PR TITLE
feat(swap): multiswap

### PR DESCRIPTION
**Pull request by ocnc**
## Summary by CodeRabbit

- **New Features**
	- Introduced a `multiSwap` function to allow for multiple sequential swaps, enhancing the swapping experience on the DEX.
	- Added a new library for defining swap details in a decentralized exchange context.

### CODE REVIEW

File: contracts/CrocSwapDex.sol [hunk 1]
```diff
@@ -93,6 +94,19 @@ contract CrocSwapDex is HotPath, ICrocMinion {
             minOut, reserveFlags, baseQuote, quoteFlow);
     }
 
+    function multiSwap (SwapHelpers.SwapStep[] memory steps) reEntrantLock public payable 
+        returns (uint128 out) {
+            require(steps.length > 0, "No steps provided");
+            require(steps[0].amount > 0, "No amount provided");
+            uint128 amountIn = steps[0].amount;
+            for (uint256 i = 0; i < steps.length; i++) {
+                (, int128 quoteFlow) = swap(steps[i].base, steps[i].quote, steps[i].poolIdx, true,
```
Comment (po-bera) [inline contracts/CrocSwapDex.sol]: Do we need a hygiene check if `step[i - 1].quote == step[i].base`?

**Comment by ocnc** [comment]
wait, my PR is bad. give me a moment

File: contracts/CrocSwapDex.sol [hunk 2]
```diff
@@ -93,6 +94,22 @@ contract CrocSwapDex is HotPath, ICrocMinion {
...
+        returns (uint128 out) {
+            require(steps.length > 0, "No steps provided");
+            require(steps[0].amount > 0, "No amount provided");
+            uint128 amountIn = steps[0].amount;
+            for (uint256 i = 0; i < steps.length; i++) {
+                // We use the max uint128 as the limit price to ensure the swap executes
+                // Given that we have full range liquidity, there is no min limit price
+                // Slippage can be controlled by the minOut parameter
+                (, int128 quoteFlow) = swap(steps[i].base, steps[i].quote, steps[i].poolIdx, true, 
+                    true, amountIn, 0, 340282366920938463463374607431768211455, steps[i].minAmountOut, 2);
```
Comment (po-bera) [inline contracts/CrocSwapDex.sol]:
```suggestion
                    true, amountIn, 0, type(uint128).max, steps[i].minAmountOut, 2);
```

**Review by po-bera — APPROVED** [review]
LGTM! @rustybera may take another look before merge?

File: contracts/CrocSwapDex.sol [hunk 3]
```diff
@@ -93,6 +93,36 @@ contract CrocSwapDex is HotPath, ICrocMinion {
...
+     * @return The token base and quote token flows associated with this swap action. 
+     *         (Negative indicates a credit paid to the user, positive a debit collected
+     *         from the user) */
+    function multiSwap (SwapHelpers.SwapStep[] memory steps) reEntrantLock public payable 
+        returns (uint128 out) {
+            require(steps.length > 0, "No steps provided");
+            require(steps[0].amount > 0, "No amount provided");
+            uint128 amountIn = steps[0].amount;
+            address baseAsset = steps[0].base;
+            for (uint256 i = 0; i < steps.length; i++) {
```
Comment (rustybera) [inline contracts/CrocSwapDex.sol]:
```suggestion
            for (uint256 i; i < steps.length; ) {
                SwapHelpers.SwapStep memory step = steps[i];
                unchecked { ++i; }
```
Replace all `steps[i]` by `step`. Gas savings.

File: contracts/CrocSwapDex.sol [hunk 4]
```diff
@@ -93,6 +93,36 @@ contract CrocSwapDex is HotPath, ICrocMinion {
...
+     *      perform.
+     *  
+     * @param steps The series of swap steps to be performed in sequence.
+     * @return The token base and quote token flows associated with this swap action. 
+     *         (Negative indicates a credit paid to the user, positive a debit collected
+     *         from the user) */
+    function multiSwap (SwapHelpers.SwapStep[] memory steps) reEntrantLock public payable 
+        returns (uint128 out) {
+            require(steps.length > 0, "No steps provided");
+            require(steps[0].amount > 0, "No amount provided");
```
Comment (rustybera) [inline contracts/CrocSwapDex.sol]:
```suggestion
            require(steps.length != 0, "No steps provided");
            require(steps[0].amount != 0, "No amount provided");
```
May be cheaper.

**Review by rustybera — APPROVED** [review]

**Review by rustybera — APPROVED** [review]

File: contracts/CrocSwapDex.sol [hunk 5]
```diff
@@ -93,6 +93,39 @@ contract CrocSwapDex is HotPath, ICrocMinion {
...
+    function multiSwap (SwapHelpers.SwapStep[] memory steps) reEntrantLock public payable 
+        returns (uint128 out) {
+            require(steps.length != 0, "No steps provided");
+            require(steps[0].amount != 0, "No amount provided");
+            uint128 amountIn = steps[0].amount;
+            address baseAsset = steps[0].base;
+            for (uint256 i; i < steps.length; ) {
+                require(steps[i].base == baseAsset, "Base asset mismatch");
+                SwapHelpers.SwapStep memory step = steps[i];
+                unchecked { ++i; }
```
Comment (rustybera) [inline contracts/CrocSwapDex.sol]:
```suggestion
                SwapHelpers.SwapStep memory step = steps[i];
                unchecked { ++i; }
                require(step.base == baseAsset, "Base asset mismatch");
```

**Review by rustybera — APPROVED** [review]

**Pull request merged**